<a href="https://colab.research.google.com/github/thomasavare/DNLP-project/blob/master/classification_DNLP_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this notebook, after realising that realizing that the model I imported (in 2 different ways) doesn't match the embeddings provided, I will try to redo the embeddings until I figure out if I can import the right SPECTER model.

----

In [ ]:
pip install -U sentence-transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 KB 6.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.8/5.8 MB 71.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 61.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.4/182.4 KB 23.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 79.1 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125938 sha256=d0d4e4e0a986dee6cb33117ae54c2e29ea6a98a28c6302196427ba459c135bc7
  Stored in directory: /root/.cache/pip/wheels/5e/6f/8c/d88aec621f3f542d26fac0342bef5e693335d125f4e54aeffe
Successfully built sentence-transformers


In [ ]:
from sentence_transformers import SentenceTransformer, util

# load model and tokenizer
model = SentenceTransformer('allenai-specter')

Downloading:   0%|          | 0.00/690 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/190 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.71k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/622 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/122 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/440M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/462k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/331 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/222k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/229 [00:00<?, ?B/s]

With the model uploaded, we can now upload the meta-data

In [ ]:
!pip install awscli
!aws s3 sync --no-sign-request s3://ai2-s2-research-public/specter/scidocs/ data/

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.0/4.0 MB 40.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 662.4/662.4 KB 51.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 73.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.6/79.6 KB 9.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 140.6/140.6 KB 15.4 MB/s eta 0:00:00
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
  Attempting uninstall: rsa
    Found existing installation: rsa 4.9
    Uninstalling rsa-4.9:
      Successfully uninstalled rsa-4.9
  Attempting uninstall: PyYAML
    Found existing installation: PyYAML 6.0
    Uninstalling PyYAML-6.0:
      Successfully uninstalled PyYAML-6.0


download: s3://ai2-s2-research-public/specter/scidocs/mag/test.csv to data/mag/test.csv
download: s3://ai2-s2-research-public/specter/scidocs/mesh/val.csv to data/mesh/val.csv
download: s3://ai2-s2-research-public/specter/scidocs/mesh/train.csv to data/mesh/train.csv
download: s3://ai2-s2-research-public/specter/scidocs/mag/val.csv to data/mag/val.csv
download: s3://ai2-s2-research-public/specter/scidocs/mesh/test.csv to data/mesh/test.csv
download: s3://ai2-s2-research-public/specter/scidocs/cite/test.qrel to data/cite/test.qrel
download: s3://ai2-s2-research-public/specter/scidocs/recomm/propensity_scores.json to data/recomm/propensity_scores.json
download: s3://ai2-s2-research-public/specter/scidocs/coread/val.qrel to data/coread/val.qrel
download: s3://ai2-s2-research-public/specter/scidocs/cocite/val.qrel to data/cocite/val.qrel
download: s3://ai2-s2-research-public/specter/scidocs/cite/val.qrel to data/cite/val.qrel
download: s3://ai2-s2-research-public/specter/scidocs/cocite/tes

In [ ]:
import pandas as pd

data_df = pd.read_json('/content/data/paper_metadata_mag_mesh.json', 'index')
data_df.columns

<ipython-input-4-c2b792037530>:3: FutureWarning: Starting with pandas version 2.0 all arguments of read_json except for the argument 'path_or_buf' will be keyword-only
  data_df = pd.read_json('/content/data/paper_metadata_mag_mesh.json', 'index')


Index(['abstract', 'authors', 'cited_by', 'paper_id', 'references', 'title',
       'year'],
      dtype='object')

We don't need the citation graph, the references and the year (maybe if my extension works).

In [ ]:
data_df = data_df[['paper_id', 'title', 'abstract']]
data_df['embedding'] = None
data_df

<ipython-input-5-94dc7cc50e51>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_df['embedding'] = None


paper_id  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9  00014a8515491f0b3fe2a1ff6e0f5305e584dcd9   
00021eeee2bf4e06fec98941206f97083c38b54d  00021eeee2bf4e06fec98941206f97083c38b54d   
00027baa2a90e1a3d50c1da14882d518de6847f5  00027baa2a90e1a3d50c1da14882d518de6847f5   
00034a5a5bd11b51ec046d31de273946d91fb766  00034a5a5bd11b51ec046d31de273946d91fb766   
000c8d85037886c86de15290e5a8e9bae7b66103  000c8d85037886c86de15290e5a8e9bae7b66103   
...                                                                            ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  ffe534bb74efd04b0038510be2a4ed5072430c8b   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66   
fff238844076ad5643dc2ff53153581bd89441ea  fff238844076ad5643dc2ff53153581bd89441ea   
fff766303c67b984a64e33566853375d35951424  fff766303c67b984a64e33566853375d35951424   

                                                                                      title  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9              ON THE CLASSIFICATION OF THE SCIENCES   
00021eeee2bf4e06fec98941206f97083c38b54d  Opportunities and challenges for E-Commerce in...   
00027baa2a90e1a3d50c1da14882d518de6847f5  Interactions between model membranes and ligni...   
00034a5a5bd11b51ec046d31de273946d91fb766        Perpetual Peace: What Kant Should Have Said   
000c8d85037886c86de15290e5a8e9bae7b66103    Reimagining Greek Tragedy on the American Stage   
...                                                                                     ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  Comparison of olmesartan combined with a calci...   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  Role of angiotensin II in plasma PAI-1 changes...   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  Factors that influence cancer patients' anxiet...   
fff238844076ad5643dc2ff53153581bd89441ea  Is adjuvant chemotherapy indicated in stage I ...   
fff766303c67b984a64e33566853375d35951424  Relationship between PCO2 and unfavorable outc...   

                                                                                   abstract  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9                                               None   
00021eeee2bf4e06fec98941206f97083c38b54d  Numerous studies of E-Commerce have emphasized...   
00027baa2a90e1a3d50c1da14882d518de6847f5  In order to elucidate the modes of interaction...   
00034a5a5bd11b51ec046d31de273946d91fb766                                               None   
000c8d85037886c86de15290e5a8e9bae7b66103  List of Illustrations Preface Introduction CHA...   
...                                                                                     ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  The cardiovascular effects of combined therapy...   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  To evaluate the relationship between plasma pl...   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  BACKGROUND\nNo study has yet assessed the impa...   
fff238844076ad5643dc2ff53153581bd89441ea  OBJECTIVE\nConservative surgery followed by pl...   
fff766303c67b984a64e33566853375d35951424  Background:Abnormal PCO2 is common in infants ...   

                                         embedding  
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9      None  
00021eeee2bf4e06fec98941206f97083c38b54d      None  
00027baa2a90e1a3d50c1da14882d518de6847f5      None  
00034a5a5bd11b51ec046d31de273946d91fb766      None  
000c8d85037886c86de15290e5a8e9bae7b66103      None  
...                                            ...  
ffe534bb74efd04b0038510be2a4ed5072430c8b      None  
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1      None  
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66      None  
fff238844076ad5643dc2ff53153581bd89441ea      None  
fff766303c67b984a64e33566853375d35951424      None  

[48473 rows x 4 columns]

Now let's try to embed everything. It may (or should) take a while to embed everything since there are 48473 rows. (takes around 3 hours, could be a good idea to save those data into a json file).

*(Here's a copy of the code after computing it)*

----
```‘{python}
from tqdm import tqdm

for index, row in tqdm(data_df.iterrows(), total=len(data_df.index)):
  data_df['embedding'][row['paper_id']] = model.encode(row['title'] + '[SEP]' + row['title'])
```
100%|██████████| 48473/48473 [3:05:07<00:00,  4.36it/s]

---

After everything was embedded, let's save those data in the cloud to not do that ever again. Will probably upload the mbeddings on my github so it's easier to import them. For now they are on my google drive which can be synchronised with the collab.

In [ ]:
embeddings_df = pd.read_json('/content/drive/MyDrive/DNLP-project/embeddings_metadata_mag_mesh.json', orient='index')

In [ ]:
embeddings_df

paper_id  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9  00014a8515491f0b3fe2a1ff6e0f5305e584dcd9   
00021eeee2bf4e06fec98941206f97083c38b54d  00021eeee2bf4e06fec98941206f97083c38b54d   
00027baa2a90e1a3d50c1da14882d518de6847f5  00027baa2a90e1a3d50c1da14882d518de6847f5   
00034a5a5bd11b51ec046d31de273946d91fb766  00034a5a5bd11b51ec046d31de273946d91fb766   
000c8d85037886c86de15290e5a8e9bae7b66103  000c8d85037886c86de15290e5a8e9bae7b66103   
...                                                                            ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  ffe534bb74efd04b0038510be2a4ed5072430c8b   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66   
fff238844076ad5643dc2ff53153581bd89441ea  fff238844076ad5643dc2ff53153581bd89441ea   
fff766303c67b984a64e33566853375d35951424  fff766303c67b984a64e33566853375d35951424   

                                                                                      title  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9              ON THE CLASSIFICATION OF THE SCIENCES   
00021eeee2bf4e06fec98941206f97083c38b54d  Opportunities and challenges for E-Commerce in...   
00027baa2a90e1a3d50c1da14882d518de6847f5  Interactions between model membranes and ligni...   
00034a5a5bd11b51ec046d31de273946d91fb766        Perpetual Peace: What Kant Should Have Said   
000c8d85037886c86de15290e5a8e9bae7b66103    Reimagining Greek Tragedy on the American Stage   
...                                                                                     ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  Comparison of olmesartan combined with a calci...   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  Role of angiotensin II in plasma PAI-1 changes...   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  Factors that influence cancer patients' anxiet...   
fff238844076ad5643dc2ff53153581bd89441ea  Is adjuvant chemotherapy indicated in stage I ...   
fff766303c67b984a64e33566853375d35951424  Relationship between PCO2 and unfavorable outc...   

                                                                                   abstract  \
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9                                               None   
00021eeee2bf4e06fec98941206f97083c38b54d  Numerous studies of E-Commerce have emphasized...   
00027baa2a90e1a3d50c1da14882d518de6847f5  In order to elucidate the modes of interaction...   
00034a5a5bd11b51ec046d31de273946d91fb766                                               None   
000c8d85037886c86de15290e5a8e9bae7b66103  List of Illustrations Preface Introduction CHA...   
...                                                                                     ...   
ffe534bb74efd04b0038510be2a4ed5072430c8b  The cardiovascular effects of combined therapy...   
ffeaac2b94fc298676e9784ed3bd7a6a7c23b9d1  To evaluate the relationship between plasma pl...   
ffecb03cbf4dccf9fcdc8b26f40f53d5bc44be66  BACKGROUND\nNo study has yet assessed the impa...   
fff238844076ad5643dc2ff53153581bd89441ea  OBJECTIVE\nConservative surgery followed by pl...   
fff766303c67b984a64e33566853375d35951424  Background:Abnormal PCO2 is common in infants ...   

                                                                                  embedding  
00014a8515491f0b3fe2a1ff6e0f5305e584dcd9  [-0.7367060184, 0.8094774485, -0.0390381552, -...  
00021eeee2bf4e06fec98941206f97083c38b54d  [-1.1841081381, 1.1308963299, 0.6682422161, 0....  
00027baa2a90e1a3d50c1da14882d518de6847f5  [0.2389703691, 0.7384790778, 1.0119105577, -0....  
00034a5a5bd11b51ec046d31de273946d91fb766  [-0.7743704319, 0.5894224048, -0.3162544966, 0...  
000c8d85037886c86de15290e5a8e9bae7b66103  [0.0163563173, 0.50049299, -0.0730320513, 0.64...  
...                                                                                     ...  
ffe534bb74efd04b0038510be2a4ed5072430c8b  [-0.5972291231, 0.17011463640000002, 0.6428002...  
ffeaac2

Now, let's import the associated classes

In [ ]:
train_path, test_path = '/content/data/mag/train.csv', '/content/data/mag/test.csv'
train_df, test_df = pd.read_csv(train_path), pd.read_csv(test_path)
train_df.columns

Index(['pid', 'class_label'], dtype='object')

In [ ]:
train_merged_df, test_merged_df = pd.merge(embeddings_df, train_df, how='inner', left_on='paper_id', right_on='pid'), pd.merge(embeddings_df, test_df, how='inner', left_on='paper_id', right_on='pid')
train_merged_df, test_merged_df = train_merged_df[['paper_id', 'embedding', 'class_label']], test_merged_df[['paper_id', 'embedding', 'class_label']]

In [ ]:
train_merged_df

paper_id  \
0      00021eeee2bf4e06fec98941206f97083c38b54d   
1      00027baa2a90e1a3d50c1da14882d518de6847f5   
2      001d134f355306faeb8993fe487a34e0ed0d278b   
3      001fb24b565bf8a5faf5551f753a55f3eed6c81c   
4      002087b2c4408d83bc095c835ce70e3974d830c7   
...                                         ...   
17496  ffeb51c4b44028bc1d5d71851f08d571c3177368   
17497  ffef3035f600d2e8cbb91192128d9315b800f71a   
17498  fff1c7f9478f1220c9f9ea341e13651cbfc2996b   
17499  fff73361b36bda4423c75f7e9a99e3996947a3c1   
17500  fffdfe2faa469ce40f44778a7d2056a88c65ac1d   

                                               embedding  class_label  
0      [-1.1841081381, 1.1308963299, 0.6682422161, 0....            2  
1      [0.2389703691, 0.7384790778, 1.0119105577, -0....            3  
2      [-1.0588871241, 0.0980567485, 0.05070229990000...           10  
3      [-0.7206891179, 0.6086265445, -0.0693209022, -...           10  
4      [0.0442002341, 0.0344819799, 0.408043295100000...            9  
...                                                  ...          ...  
17496  [-0.3375469446, 0.5189490914, -0.2140062898, -...           12  
17497  [-0.33342462780000004, 0.8680065274000001, 0.2...           17  
17498  [-1.2471318245, 0.6116715074, 0.36738637090000...           15  
17499  [-1.0980789661, 0.3576785326, -0.2542884648, -...           16  
17500  [-1.5524930954, 0.9185175896000001, 0.18831063...            3  

[17501 rows x 3 columns]

Time to train a classifier! 

First, let's prepare the data into an adequat object.

In [ ]:
import numpy as np

x_train, y_train = [], []
x_test, y_test = [], []

i = 0
for _, embedding, class_label in train_merged_df.values:
  x_train.append(embedding)
  y_train.append(class_label)

x_train = np.array(x_train)
y_train = np.array(y_train)

for _, embedding, class_label in test_merged_df.values:
  x_test.append(embedding)
  y_test.append(class_label)

x_test = np.array(x_test)
y_test = np.array(y_test)

In [ ]:
from sklearn import svm

classifier1 = svm.LinearSVC(penalty='l2', loss='squared_hinge', dual=True, C=1.0, multi_class='ovr', random_state=42)

In [ ]:
classifier1.fit(x_train, y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


LinearSVC(random_state=42)

In [ ]:
from sklearn.metrics import f1_score

y_pred = classifier1.predict(x_test)
classifier1.score(x_test, y_test), f1_score(y_test, y_pred, average='macro')

(0.7206078379098907, 0.7105471368841444)

observations on f1 score

In [ ]:
def classify_article(classifier, title, abstract, cls=False, probas=False):
  classes = {0: "Art", 1: "Biology", 2: "Business", 3: "Chemistry", 4: "Computer science", 5: "Economics", 6: "Engineering", 7: "Environmental science",\
             8: "Geography", 9: "Geology", 10: "History", 11: "Materials science", 12: "Mathematics", 13: "Medicine", 14: "Philosophy", 15: "Physics",\
             16: "Political science", 17: "Psychology", 18: "Sociology"}

  embed = model.encode([title + '[SEP]' + abstract])

  if cls:
    return np.array(list(zip(classifier.decision_function(embed)[0], classes.values())), dtype=object)
  if probas:
    return np.array(list(zip(classifier.predict_proba(embed)[0], classes.values())), dtype=object)
  return classes[classifier.decision_function(embed)]

Title: SPECTER: Document-level Representation Learning using Citation-informed Transformers

Abstract: Representation learning is a critical ingredient for natural language processing systems. Recent Transformer language models like BERT learn powerful textual representations, but these models are targeted towards token- and sentence-level training objectives and do not leverage information on inter-document relatedness, which limits their document-level representation power. For applications on scientific documents, such as classification and recommendation, the embeddings power strong performance on end tasks. We propose SPECTER, a new method to generate document-level embedding of scientific documents based on pretraining a Transformer language model on a powerful signal of document-level relatedness: the citation graph. Unlike existing pretrained language models, SPECTER can be easily applied to downstream applications without task-specific fine-tuning. Additionally, to encourage further research on document-level models, we introduce SCIDOCS, a new evaluation benchmark consisting of seven document-level tasks ranging from citation prediction, to document classification and recommendation. We show that SPECTER outperforms a variety of competitive baselines on the benchmark.

My Guess: Computer science (obviously)


In [ ]:
res = classify_article(classifier1, "SPECTER: Document-level Representation Learning using Citation-informed Transformers", "Representation learning is a critical ingredient for natural language processing systems. Recent Transformer language models like BERT learn powerful textual representations, but these models are targeted towards token- and sentence-level training objectives and do not leverage information on inter-document relatedness, which limits their document-level representation power. For applications on scientific documents, such as classification and recommendation, the embeddings power strong performance on end tasks. We propose SPECTER, a new method to generate document-level embedding of scientific documents based on pretraining a Transformer language model on a powerful signal of document-level relatedness: the citation graph. Unlike existing pretrained language models, SPECTER can be easily applied to downstream applications without task-specific fine-tuning. Additionally, to encourage further research on document-level models, we introduce SCIDOCS, a new evaluation benchmark consisting of seven document-level tasks ranging from citation prediction, to document classification and recommendation. We show that SPECTER outperforms a variety of competitive baselines on the benchmark.", cls=True)
res[np.argsort(-res[:, 0])]

array([[3.2369296722674905, 'Computer science'],
       [-1.1857062291402207, 'Geography'],
       [-1.716168860218461, 'Political science'],
       [-2.1038272318817426, 'Economics'],
       [-2.2295005016877956, 'Art'],
       [-2.291139931264881, 'Sociology'],
       [-2.4857453916888557, 'Mathematics'],
       [-2.565927137039865, 'History'],
       [-3.7247929432156073, 'Environmental science'],
       [-3.7729087297188744, 'Psychology'],
       [-4.368896558508119, 'Chemistry'],
       [-4.412378605587704, 'Philosophy'],
       [-4.70736910432564, 'Engineering'],
       [-4.943433695584021, 'Business'],
       [-5.4857111370386535, 'Geology'],
       [-5.640873947151579, 'Materials science'],
       [-6.079235584645333, 'Physics'],
       [-6.795215135272074, 'Biology'],
       [-7.547455029785584, 'Medicine']], dtype=object)

Title: Satellite Content and Halo Mass of Galaxy Clusters: Comparison between Red-Sequence and Halo-based Optical Cluster Finders

Abstract: Cluster cosmology depends critically on how the optical clusters are selected from imaging surveys. We compare the conditional luminosity function (CLF) and weak lensing halo masses between two different cluster samples at fixed richness, detected within the same volume (0.1<z<0.34) using the red-sequence and halo-based methods. After calibrating our CLF deprojection method against mock galaxy samples, we measure the 3D CLFs by cross-correlating clusters with SDSS photometric galaxies. As expected, the CLFs of the red-sequence and halo-based finders exhibit redder and bluer populations, respectively. We also find significant shape discrepancies between the two CLFs at the faint end, where the red-sequence clusters show a strong deficit of faint galaxies but a bump at Mr∼−20.5, while the halo-based clusters host an increasing number of blue satellites. By comparing the subsamples of clusters that have a match between the two catalogues to those without matches, we discover that the CLF shape depends sensitively on the cluster centroiding. However, the average weak lensing halo mass between the matched and non-matched clusters are roughly consistent with each other in either cluster sample. Since the colour preferences of the two cluster finders are almost orthogonal, such a consistency indicates that the scatter in the mass-richness relation of either cluster sample is close to random. Therefore, while the choice of how optical clusters are identified impacts the satellite content, our result suggests that it should not introduce strong systematics biases in cluster cosmology.

Guess: In (astro) physics category on arxiv.org

In [ ]:
res = classify_article(classifier1, "Satellite Content and Halo Mass of Galaxy Clusters: Comparison between Red-Sequence and Halo-based Optical Cluster Finders", "Cluster cosmology depends critically on how the optical clusters are selected from imaging surveys. We compare the conditional luminosity function (CLF) and weak lensing halo masses between two different cluster samples at fixed richness, detected within the same volume (0.1<z<0.34) using the red-sequence and halo-based methods. After calibrating our CLF deprojection method against mock galaxy samples, we measure the 3D CLFs by cross-correlating clusters with SDSS photometric galaxies. As expected, the CLFs of the red-sequence and halo-based finders exhibit redder and bluer populations, respectively. We also find significant shape discrepancies between the two CLFs at the faint end, where the red-sequence clusters show a strong deficit of faint galaxies but a bump at Mr∼−20.5, while the halo-based clusters host an increasing number of blue satellites. By comparing the subsamples of clusters that have a match between the two catalogues to those without matches, we discover that the CLF shape depends sensitively on the cluster centroiding. However, the average weak lensing halo mass between the matched and non-matched clusters are roughly consistent with each other in either cluster sample. Since the colour preferences of the two cluster finders are almost orthogonal, such a consistency indicates that the scatter in the mass-richness relation of either cluster sample is close to random. Therefore, while the choice of how optical clusters are identified impacts the satellite content, our result suggests that it should not introduce strong systematics biases in cluster cosmology.", cls=True)
res[np.argsort(-res[:, 0])]

array([[-0.7727859468382251, 'Computer science'],
       [-1.1091047287947644, 'Chemistry'],
       [-1.7760194008620345, 'Sociology'],
       [-1.8185303820598646, 'Psychology'],
       [-2.2373288384007033, 'Geography'],
       [-3.2813099254393086, 'Political science'],
       [-4.309172941167755, 'Business'],
       [-4.469313478096465, 'Economics'],
       [-5.084784306353317, 'History'],
       [-5.166778983709949, 'Physics'],
       [-5.461507029251253, 'Geology'],
       [-5.497355607619488, 'Philosophy'],
       [-6.132283828410673, 'Engineering'],
       [-6.215895815659463, 'Art'],
       [-6.2404871733729745, 'Mathematics'],
       [-6.379595281416484, 'Materials science'],
       [-6.896681616481737, 'Medicine'],
       [-8.57880849496606, 'Environmental science'],
       [-8.768306448264143, 'Biology']], dtype=object)

title: Exploring The Concept Of Cubism Art Essay

abstract: Cubism was one of the most influential art movements of the 20th century. It took place between 1907 and about 1914. The innovators of the Cubist movement were Pablo Picasso (Spanish, 1881 1973) and Georges Braque (French, 1882 1963). Cubism was one of the most significant changes in ideas in the history of art. It allowed for the development of many of the abstract modern art movements in areas such as Futurism and Constructivism.

In [ ]:
res = classify_article(classifier1, "Exploring The Concept Of Cubism Art Essay", "Cubism was one of the most influential art movements of the 20th century. It took place between 1907 and about 1914. The innovators of the Cubist movement were Pablo Picasso (Spanish, 1881 1973) and Georges Braque (French, 1882 1963). Cubism was one of the most significant changes in ideas in the history of art. It allowed for the development of many of the abstract modern art movements in areas such as Futurism and Constructivism.", cls=True)
res[np.argsort(-res[:, 0])]

array([[0.6367172470143455, 'Art'],
       [-0.02327048836533857, 'Sociology'],
       [-0.23059300416444772, 'Mathematics'],
       [-0.32591475917339363, 'Economics'],
       [-0.8426298229210223, 'History'],
       [-1.7972712431993418, 'Political science'],
       [-2.0760977984100877, 'Philosophy'],
       [-2.2895093194718465, 'Computer science'],
       [-2.4575364442189374, 'Geography'],
       [-2.5867551257263033, 'Psychology'],
       [-3.1378339909753725, 'Physics'],
       [-4.032798679854295, 'Biology'],
       [-4.411666233356051, 'Engineering'],
       [-5.287606310997519, 'Geology'],
       [-5.557289380822116, 'Materials science'],
       [-5.78371795439588, 'Medicine'],
       [-6.7180000353912765, 'Business'],
       [-8.08379126360974, 'Chemistry'],
       [-8.93821187993181, 'Environmental science']], dtype=object)

Seems to work, I was just stupid and stubborn for 1 month T$\cap$T.

----
Let's try another type of SVM.

In [ ]:
classifier2 = svm.SVC(decision_function_shape='ovo', gamma='auto', probability=True)

In [ ]:
classifier2.fit(x_train, y_train)

SVC(decision_function_shape='ovo', gamma='auto', probability=True)

In [ ]:
classifier2.decision_function_shape = 'ovr'

In [ ]:
y_pred = classifier2.predict(x_test)

In [ ]:
y_pred = classifier2.predict(x_test)
f1_score(y_test, y_pred, average='macro'), classifier1.score(x_test, y_test)

(0.7411100630272639, 0.7206078379098907)

improvment in f1_score (and accuracy). That's cool. Let's put it to the test.

In [ ]:
res = classify_article(classifier2, "SPECTER: Document-level Representation Learning using Citation-informed Transformers", "Representation learning is a critical ingredient for natural language processing systems. Recent Transformer language models like BERT learn powerful textual representations, but these models are targeted towards token- and sentence-level training objectives and do not leverage information on inter-document relatedness, which limits their document-level representation power. For applications on scientific documents, such as classification and recommendation, the embeddings power strong performance on end tasks. We propose SPECTER, a new method to generate document-level embedding of scientific documents based on pretraining a Transformer language model on a powerful signal of document-level relatedness: the citation graph. Unlike existing pretrained language models, SPECTER can be easily applied to downstream applications without task-specific fine-tuning. Additionally, to encourage further research on document-level models, we introduce SCIDOCS, a new evaluation benchmark consisting of seven document-level tasks ranging from citation prediction, to document classification and recommendation. We show that SPECTER outperforms a variety of competitive baselines on the benchmark.", probas=True)
res[np.argsort(-res[:, 0])]

array([[0.9414180906486594, 'Computer science'],
       [0.013398300384947557, 'Engineering'],
       [0.008113647627665506, 'Mathematics'],
       [0.004707301479023342, 'Political science'],
       [0.004464943203951814, 'Art'],
       [0.003682868813241793, 'Business'],
       [0.0030409946300757095, 'Economics'],
       [0.0026173170751954194, 'History'],
       [0.0025817533913157185, 'Materials science'],
       [0.0024234815769820215, 'Geology'],
       [0.0023968089285539337, 'Geography'],
       [0.002283664801887463, 'Psychology'],
       [0.002093192702209119, 'Sociology'],
       [0.0019688918903567124, 'Environmental science'],
       [0.0012247823308444102, 'Biology'],
       [0.0011362852186324963, 'Chemistry'],
       [0.0009833328113286394, 'Philosophy'],
       [0.0007580982110527078, 'Medicine'],
       [0.0007062442740759331, 'Physics']], dtype=object)

In [ ]:
res = classify_article(classifier2, "Satellite Content and Halo Mass of Galaxy Clusters: Comparison between Red-Sequence and Halo-based Optical Cluster Finders", "Cluster cosmology depends critically on how the optical clusters are selected from imaging surveys. We compare the conditional luminosity function (CLF) and weak lensing halo masses between two different cluster samples at fixed richness, detected within the same volume (0.1<z<0.34) using the red-sequence and halo-based methods. After calibrating our CLF deprojection method against mock galaxy samples, we measure the 3D CLFs by cross-correlating clusters with SDSS photometric galaxies. As expected, the CLFs of the red-sequence and halo-based finders exhibit redder and bluer populations, respectively. We also find significant shape discrepancies between the two CLFs at the faint end, where the red-sequence clusters show a strong deficit of faint galaxies but a bump at Mr∼−20.5, while the halo-based clusters host an increasing number of blue satellites. By comparing the subsamples of clusters that have a match between the two catalogues to those without matches, we discover that the CLF shape depends sensitively on the cluster centroiding. However, the average weak lensing halo mass between the matched and non-matched clusters are roughly consistent with each other in either cluster sample. Since the colour preferences of the two cluster finders are almost orthogonal, such a consistency indicates that the scatter in the mass-richness relation of either cluster sample is close to random. Therefore, while the choice of how optical clusters are identified impacts the satellite content, our result suggests that it should not introduce strong systematics biases in cluster cosmology.", probas=True)
res[np.argsort(-res[:, 0])]

array([[0.8054168613860793, 'Physics'],
       [0.05638272157718863, 'Geology'],
       [0.04858051337894907, 'Chemistry'],
       [0.018454565022609566, 'Materials science'],
       [0.014645814274950479, 'Computer science'],
       [0.011152072111711679, 'Biology'],
       [0.008288149539953705, 'Geography'],
       [0.006292100411289724, 'Business'],
       [0.005087706232878692, 'Mathematics'],
       [0.00407907398975628, 'Environmental science'],
       [0.004058251054174858, 'Psychology'],
       [0.003363495834398129, 'Economics'],
       [0.0026722461861587435, 'Political science'],
       [0.0026592543026966887, 'Engineering'],
       [0.0021444590041127375, 'Medicine'],
       [0.002032932156248615, 'Sociology'],
       [0.0017784252446236188, 'Art'],
       [0.0016782824491071457, 'History'],
       [0.001233075843111124, 'Philosophy']], dtype=object)

In [ ]:
res = classify_article(classifier2, "Exploring The Concept Of Cubism Art Essay", "Cubism was one of the most influential art movements of the 20th century. It took place between 1907 and about 1914. The innovators of the Cubist movement were Pablo Picasso (Spanish, 1881 1973) and Georges Braque (French, 1882 1963). Cubism was one of the most significant changes in ideas in the history of art. It allowed for the development of many of the abstract modern art movements in areas such as Futurism and Constructivism.", probas=True)
res[np.argsort(-res[:, 0])]

array([[0.8924496490859399, 'Art'],
       [0.04285157902996371, 'History'],
       [0.027243813908463133, 'Sociology'],
       [0.010343518506481401, 'Philosophy'],
       [0.005980852234985908, 'Political science'],
       [0.00539430121055874, 'Engineering'],
       [0.0035952243976042604, 'Geography'],
       [0.0027933726663202565, 'Economics'],
       [0.00266307121360342, 'Psychology'],
       [0.002173799986851709, 'Computer science'],
       [0.001745330200741101, 'Geology'],
       [0.0007432245869118252, 'Mathematics'],
       [0.000604081301904381, 'Business'],
       [0.0003827639420842821, 'Materials science'],
       [0.0003604745978121696, 'Physics'],
       [0.00022711096748678704, 'Chemistry'],
       [0.00019412583590743, 'Medicine'],
       [0.0001323063489452074, 'Environmental science'],
       [0.00012139997743512722, 'Biology']], dtype=object)

----

Let's try a simple CNN

In [ ]:
import tensorflow as tf
y = tf.keras.utils.to_categorical(y_train)

In [ ]:
from tensorflow.keras import layers

inputs = tf.keras.Input(shape=(x_train.shape[1], 1,), dtype="float")

x = layers.Conv1D(128, 1)(inputs)
x = layers.Conv1D(64, 2, strides=2)(x)
x = layers.AveragePooling1D(strides=2)(x)
x = layers.Conv1D(64, 2)(x)
x = layers.AveragePooling1D(strides=2)(x)
x = layers.Flatten()(x)
# We add a vanilla hidden layer:
x = layers.Dense(64, activation="relu")(x)
x = layers.Dense(32, activation="relu")(x)

# We project onto a single unit output layer, and squash it with a sigmoid:
predictions = layers.Dense(19, activation="sigmoid", name="predictions")(x)

classifier = tf.keras.Model(inputs, predictions)

In [ ]:
classifier.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 768, 1)]          0         
                                                                 
 conv1d (Conv1D)             (None, 768, 128)          256       
                                                                 
 conv1d_1 (Conv1D)           (None, 384, 64)           16448     
                                                                 
 average_pooling1d (AverageP  (None, 192, 64)          0         
 ooling1D)                                                       
                                                                 
 conv1d_2 (Conv1D)           (None, 191, 64)           8256      
                                                                 
 average_pooling1d_1 (Averag  (None, 95, 64)           0         
 ePooling1D)                                                 

In [ ]:
from keras import backend as K

def recall_m(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    recall = true_positives / (possible_positives + K.epsilon())
    return recall

def precision_m(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    precision = true_positives / (predicted_positives + K.epsilon())
    return precision

def f1_m(y_true, y_pred):
    precision = precision_m(y_true, y_pred)
    recall = recall_m(y_true, y_pred)
    return 2*((precision*recall)/(precision+recall+K.epsilon()))

In [ ]:
classifier.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy", f1_m])

will try squared hinge loss later maybe.

In [ ]:
x_train_cnn = np.expand_dims(x_train, -1)
x_train_cnn.shape

(17501, 768, 1)

In [ ]:
epochs = 20

# Fit the model using the train and test datasets.
classifier.fit(x_train_cnn, y, epochs=epochs)

Epoch 1/20
547/547 [==============================] - 49s 87ms/step - loss: 1.0920 - accuracy: 0.6536 - f1_m: 0.1903
Epoch 2/20
547/547 [==============================] - 42s 77ms/step - loss: 0.8584 - accuracy: 0.7180 - f1_m: 0.2125
Epoch 3/20
547/547 [==============================] - 44s 80ms/step - loss: 0.8065 - accuracy: 0.7334 - f1_m: 0.2205
Epoch 4/20
547/547 [==============================] - 44s 80ms/step - loss: 0.7723 - accuracy: 0.7437 - f1_m: 0.2284
Epoch 5/20
547/547 [==============================] - 44s 80ms/step - loss: 0.7487 - accuracy: 0.7505 - f1_m: 0.2352
Epoch 6/20
547/547 [==============================] - 44s 80ms/step - loss: 0.7275 - accuracy: 0.7564 - f1_m: 0.2387
Epoch 7/20
547/547 [==============================] - 44s 80ms/step - loss: 0.7121 - accuracy: 0.7600 - f1_m: 0.2469
Epoch 8/20
547/547 [==============================] - 44s 81ms/step - loss: 0.6935 - accuracy: 0.7656 - f1_m: 0.2524
Epoch 9/20
547/547 [==============================] - 44s 80ms/s

In [ ]:
y_test_cnn = tf.keras.utils.to_categorical(y_test)
score = classifier.evaluate(x_test, y_test_cnn)
score

118/118 [==============================] - 3s 23ms/step - loss: 0.9216 - accuracy: 0.7193 - f1_m: 0.2856


[0.9216019511222839, 0.7192748785018921, 0.28564757108688354]

wrong computation of f1, will correct it if I have the time to.

In [ ]:
y_pred = classifier.predict(x_test)
f1_score(y_test, np.argmax(y_pred, axis=1), average='macro')

118/118 [==============================] - 3s 22ms/step


0.7065101041769221

In [ ]:
def classify_article2(title, abstract, cls=False):
  classes = {0: "Art", 1: "Biology", 2: "Business", 3: "Chemistry", 4: "Computer science", 5: "Economics", 6: "Engineering", 7: "Environmental science",\
             8: "Geography", 9: "Geology", 10: "History", 11: "Materials science", 12: "Mathematics", 13: "Medicine", 14: "Philosophy", 15: "Physics",\
             16: "Political science", 17: "Psychology", 18: "Sociology"}
   
  if cls:
    return np.array(list(zip(classifier.predict(model.encode([title + '[SEP]' + abstract]))[0], classes.values())), dtype=object)
  return classes[np.argmax(classifier.predict(model.encode([title + '[SEP]' + abstract]))[0])]

In [ ]:
res = classify_article2("SPECTER: Document-level Representation Learning using Citation-informed Transformers", "Representation learning is a critical ingredient for natural language processing systems. Recent Transformer language models like BERT learn powerful textual representations, but these models are targeted towards token- and sentence-level training objectives and do not leverage information on inter-document relatedness, which limits their document-level representation power. For applications on scientific documents, such as classification and recommendation, the embeddings power strong performance on end tasks. We propose SPECTER, a new method to generate document-level embedding of scientific documents based on pretraining a Transformer language model on a powerful signal of document-level relatedness: the citation graph. Unlike existing pretrained language models, SPECTER can be easily applied to downstream applications without task-specific fine-tuning. Additionally, to encourage further research on document-level models, we introduce SCIDOCS, a new evaluation benchmark consisting of seven document-level tasks ranging from citation prediction, to document classification and recommendation. We show that SPECTER outperforms a variety of competitive baselines on the benchmark.", cls=True)
res[np.argsort(-res[:, 0])]

1/1 [==============================] - 0s 18ms/step


array([[0.9999999, 'Computer science'],
       [0.99414927, 'Mathematics'],
       [0.88806415, 'Environmental science'],
       [0.81128025, 'Engineering'],
       [0.49066144, 'Geography'],
       [0.2700923, 'Geology'],
       [0.06653451, 'Physics'],
       [0.022941021, 'Psychology'],
       [0.014344574, 'Materials science'],
       [0.008268426, 'Business'],
       [0.0058662225, 'History'],
       [0.0037175068, 'Political science'],
       [0.0033495498, 'Philosophy'],
       [0.0022523724, 'Sociology'],
       [0.0020437972, 'Economics'],
       [0.001682808, 'Biology'],
       [0.00048200408, 'Chemistry'],
       [0.00040793794, 'Art'],
       [8.1909465e-11, 'Medicine']], dtype=object)

In [ ]:
res = classify_article2("Satellite Content and Halo Mass of Galaxy Clusters: Comparison between Red-Sequence and Halo-based Optical Cluster Finders", "Cluster cosmology depends critically on how the optical clusters are selected from imaging surveys. We compare the conditional luminosity function (CLF) and weak lensing halo masses between two different cluster samples at fixed richness, detected within the same volume (0.1<z<0.34) using the red-sequence and halo-based methods. After calibrating our CLF deprojection method against mock galaxy samples, we measure the 3D CLFs by cross-correlating clusters with SDSS photometric galaxies. As expected, the CLFs of the red-sequence and halo-based finders exhibit redder and bluer populations, respectively. We also find significant shape discrepancies between the two CLFs at the faint end, where the red-sequence clusters show a strong deficit of faint galaxies but a bump at Mr∼−20.5, while the halo-based clusters host an increasing number of blue satellites. By comparing the subsamples of clusters that have a match between the two catalogues to those without matches, we discover that the CLF shape depends sensitively on the cluster centroiding. However, the average weak lensing halo mass between the matched and non-matched clusters are roughly consistent with each other in either cluster sample. Since the colour preferences of the two cluster finders are almost orthogonal, such a consistency indicates that the scatter in the mass-richness relation of either cluster sample is close to random. Therefore, while the choice of how optical clusters are identified impacts the satellite content, our result suggests that it should not introduce strong systematics biases in cluster cosmology.", cls=True)
res[np.argsort(-res[:, 0])]

1/1 [==============================] - 0s 24ms/step


array([[1.0, 'Physics'],
       [0.99999934, 'Chemistry'],
       [0.9982926, 'Mathematics'],
       [0.99774337, 'Geology'],
       [0.9793127, 'Materials science'],
       [0.9716356, 'Geography'],
       [0.81944937, 'Business'],
       [0.8074698, 'Computer science'],
       [0.7683715, 'Environmental science'],
       [0.1989159, 'Philosophy'],
       [0.19356726, 'Psychology'],
       [0.025063658, 'Economics'],
       [0.021586807, 'Biology'],
       [0.020153387, 'Sociology'],
       [0.008896165, 'History'],
       [0.002268006, 'Political science'],
       [0.0021029266, 'Engineering'],
       [9.382865e-09, 'Art'],
       [2.3495374e-13, 'Medicine']], dtype=object)

In [ ]:
res = classify_article2("Exploring The Concept Of Cubism Art Essay", "Cubism was one of the most influential art movements of the 20th century. It took place between 1907 and about 1914. The innovators of the Cubist movement were Pablo Picasso (Spanish, 1881 1973) and Georges Braque (French, 1882 1963). Cubism was one of the most significant changes in ideas in the history of art. It allowed for the development of many of the abstract modern art movements in areas such as Futurism and Constructivism.", cls=True)
res[np.argsort(-res[:, 0])]

1/1 [==============================] - 0s 18ms/step


array([[0.999854, 'Art'],
       [0.9906299, 'History'],
       [0.9493535, 'Sociology'],
       [0.8793125, 'Philosophy'],
       [0.7558811, 'Political science'],
       [0.6130978, 'Psychology'],
       [0.47238475, 'Geography'],
       [0.28493187, 'Engineering'],
       [0.23878404, 'Economics'],
       [0.020241622, 'Computer science'],
       [0.017087078, 'Mathematics'],
       [0.01313262, 'Physics'],
       [0.0040359437, 'Geology'],
       [0.0029771705, 'Biology'],
       [0.00079691375, 'Environmental science'],
       [0.0002191038, 'Medicine'],
       [0.00013795015, 'Materials science'],
       [0.00013295913, 'Business'],
       [1.712118e-06, 'Chemistry']], dtype=object)

----

Right now, all the performances are much lower than the one benchmarked but will try later to match them.

Here we're trying to reproduce the benchmark model according to the scidocs github.

The gridsearch allows to find the model with the best parameters. Unlilke scidoc, we're also searching for regular svm with an rbf kernel and not only linear SVC (might take some time, about an 1h).

In [ ]:
from sklearn.model_selection import GridSearchCV

parameters = {'kernel':('linear', 'rbf'), 'C': np.logspace(-4, 2, 7)}
svc = svm.SVC()
clf = GridSearchCV(svc, parameters)

In [ ]:
#clf.fit(x_train, y_train)

In [ ]:
#y_pred = clf.predict(x_test)
#clf.score(x_test, y_test), f1_score(y_test, y_pred, average='macro')

In [ ]:
#clf.get_params()

So after the grid search, we end up with the same parameters as our first classifier (wow that was a very useful wait T$\cap$T) but with a little better f1 score than before. We are still 7 points under the scidocs benchmark on f1 score.

----

Let's try other models.

First we're going to try an SGD classifier with a  a square hinge loss (the normal hinge is just a linear SVM), perceptron loss.

In [ ]:
from sklearn.linear_model import SGDClassifier

parameters = {'loss': ('perceptron', 'squared_hinge')} 
sgd = SGDClassifier()
clf2 = GridSearchCV(sgd, parameters)

In [ ]:
clf2.fit(x_train, y_train)

GridSearchCV(estimator=SGDClassifier(),
             param_grid={'loss': ('perceptron', 'squared_hinge')})

In [ ]:
clf2.get_params()

{'cv': None,
 'error_score': nan,
 'estimator__alpha': 0.0001,
 'estimator__average': False,
 'estimator__class_weight': None,
 'estimator__early_stopping': False,
 'estimator__epsilon': 0.1,
 'estimator__eta0': 0.0,
 'estimator__fit_intercept': True,
 'estimator__l1_ratio': 0.15,
 'estimator__learning_rate': 'optimal',
 'estimator__loss': 'hinge',
 'estimator__max_iter': 1000,
 'estimator__n_iter_no_change': 5,
 'estimator__n_jobs': None,
 'estimator__penalty': 'l2',
 'estimator__power_t': 0.5,
 'estimator__random_state': None,
 'estimator__shuffle': True,
 'estimator__tol': 0.001,
 'estimator__validation_fraction': 0.1,
 'estimator__verbose': 0,
 'estimator__warm_start': False,
 'estimator': SGDClassifier(),
 'n_jobs': None,
 'param_grid': {'loss': ('perceptron', 'squared_hinge')},
 'pre_dispatch': '2*n_jobs',
 'refit': True,
 'return_train_score': False,
 'scoring': None,
 'verbose': 0}

In [ ]:
y_pred = clf2.predict(x_test)
clf2.score(x_test, y_test), f1_score(y_test, y_pred, average='macro')

(0.7014129565449213, 0.6880289729080052)